File: libs/src/typescript/salmon-fs/salmonfs/drive/drive_key.ts
/*
MIT License

Copyright (c) 2021 Max Kas

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
*/

/**
 * Encryption keys and properties.
 */
export class DriveKey {

    #masterKey: Uint8Array | null = null;
    #driveKey: Uint8Array | null = null;
    #hashKey: Uint8Array | null = null;
    #iterations: number = 0;

    /**
     * Clear the properties from memory.
     */
    public clear(): void {

        if (this.#driveKey)
            this.#driveKey.fill(0);
        this.#driveKey = null;

        if (this.#hashKey)
            this.#hashKey.fill(0);
        this.#hashKey = null;

        if (this.#masterKey)
            this.#masterKey.fill(0);
        this.#masterKey = null;

        this.#iterations = 0;
    }

    /**
     * Function returns the encryption key that will be used to encrypt/decrypt the files
     * @returns {Uint8Array | null} The drive key
     */
    public getDriveKey(): Uint8Array | null {
        return this.#driveKey;
    }

    /**
     * Function returns the hash key that will be used to sign the file chunks
     * @returns {Uint8Array | null} The hash key
     */
    public getHashKey(): Uint8Array | null {
        return this.#hashKey;
    }

    /**
     * Set the drive key.
     * @param {Uint8Array | null} driveKey The drive key
     */
    public setDriveKey(driveKey: Uint8Array | null): void {
        this.#driveKey = driveKey;
    }

    /**
     * Set the hash key.
     * @param {Uint8Array | null} hashKey The hash key
     */
    public setHashKey(hashKey: Uint8Array | null): void {
        this.#hashKey = hashKey;
    }

    /**
     * Get the master key.
     * @return {Uint8Array | null} The master key
     */
    public getMasterKey(): Uint8Array | null {
        return this.#masterKey;
    }

    /**
     * Set the master key.
     * @param {Uint8Array | null} masterKey The master key
     */
    public setMasterKey(masterKey: Uint8Array | null): void {
        this.#masterKey = masterKey;
    }

    /**
     * Get the number of iterations for the master key derivation.
     * @return {number} The iterations
     */
    public getIterations(): number {
        return this.#iterations;
    }

    /**
     * Set the number of iterations for the master key derivation.
     * @param {number} iterations The iterations
     */
    public setIterations(iterations: number): void {
        this.#iterations = iterations;
    }
}

File: libs/src/python/salmon_fs/salmon_fs/salmonfs/drive/drive_key.py
#!/usr/bin/env python3
__license__ = """
MIT License

Copyright (c) 2021 Max Kas

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
"""

from typeguard import typechecked


@typechecked
class DriveKey:
    """
    Encryption keys and properties.
    """

    def __init__(self):
        self.__masterKey: bytearray | None = None
        self.__driveKey: bytearray | None = None
        self.__hash_key: bytearray | None = None
        self.__iterations: int = 0

    def clear(self):
        """
        Clear the properties from memory.
        """
        if self.__driveKey is not None:
            self.__driveKey[0:len(self.__driveKey)] = [0] * len(self.__driveKey)
        self.__driveKey = None

        if self.__hash_key is not None:
            self.__hash_key[0:len(self.__hash_key)] = [0] * len(self.__hash_key)
        self.__hash_key = None

        if self.__masterKey is not None:
            self.__masterKey[0:len(self.__masterKey)] = [0] * len(self.__masterKey)
        self.__masterKey = None
        self.__iterations = 0

    def get_drive_key(self) -> bytearray | None:
        """
        Function returns the encryption key that will be used to encrypt/decrypt the files
        """
        return self.__driveKey

    def get_hash_key(self) -> bytearray | None:
        """
        Function returns the hash key that will be used to sign the file chunks
        """
        return self.__hash_key

    def set_drive_key(self, drive_key: bytearray):
        """
        Set the drive key.
        :param drive_key:         """
        self.__driveKey = drive_key

    def set_hash_key(self, hash_key: bytearray):
        """
        Set the hash key.
        :param hash_key:         """
        self.__hash_key = hash_key

    def get_master_key(self) -> bytearray:
        """
        Get the master key.
        :return: The master key
        """
        return self.__masterKey

    def set_master_key(self, master_key: bytearray):
        """
        Set the master key.
        :param master_key:         """
        self.__masterKey = master_key

    def get_iterations(self) -> int:
        """
        Get the number of iterations for the master key derivation.
        :return: The iterations
        """
        return self.__iterations

    def set_iterations(self, iterations: int):
        """
        Set the number of iterations for the master key derivation.
        :param iterations:         """
        self.__iterations = iterations
